lib/pbench/server/cache_manager.py

```diff
@@ -226,6 +238,30 @@ def create(cls, tarball: Path, controller: "Controller") -> "Tarball":
 
         return cls(destination, controller)
 
+    def cache_map(self, tarball: Path):
+        """Builds Hierarchy structure of a Tarball in a Dictionary
+        Format.
+
+        Args:
+            tarball: path of an unpacked tarball
+        """
+        cmap = {tarball.name: {"details": TarInfo(tarball)}}
+        dir_queue = [(tarball, cmap)]
+        while dir_queue:
+            tar_path, parent_map = dir_queue.pop(0)
+            tar_n = tar_path.name
+            curr = parent_map[tar_n]
+
+            dir_list = tar_path.glob("*")
```

Comment:
Using the `tar` command is a *lot* faster than using the Python package, especially for large tarballs. This first iteration, however, is simply building a map of what we've already unpacked. I think that, as we move forward, we'll want to base this off a `tar t` rather than really unpacking anything and simply building the map so that we can extract individual files or subtrees as we need them, then keep track of where they are, how old they are, and how recently (maybe how much) they've been used.

Building the cache map structure is the first step, and doing that on top of the (temporary) unpack directory makes as much sense as anything. (Right now the indexer is going to unpack everything anyway, so there's no reason to dig into the tarball yet another time.)

Reply:
> Using the tar command is a lot faster than using the Python package

"A lot faster" for what?  In this blob of code, all we really want is a list of file paths.  We don't need to unpack the tarball, and we just need to scan through it once.  Now, it's possible, even for that, that the `tar(1)` command is faster than the Python library package, but, given that we're not actually writing anything to the file system, I bet it would be hard to tell.  However, if you are going to compare the operation of the Python package, not to the `tar(1)` command, but to a file system already containing an unpacked tarball, I would be more than willing to believe that traversing the file system is faster than reading a tarball...but, that is a possibly unfair comparison, since I'm hoping that we're going to get to where we skip the unpacking, which means we won't have the file system laid out for us to read, so we won't get the advantage of already being at the finish line when race starts.

But, yes, as an evolutionary step, since we already have the tarball unpacked, it's not unreasonable to build the cache map from the file system, but, as you said earlier, we need to guard against developing dependencies on artifacts of this implementation (like pulling the last-modified timestamp from it).  The question is, if we're not going to have the file system to depend on in the future, should we stop using it now?  Perhaps; perhaps not....

Reply:
OK, so I'd previously done a comparison of extract times for `tarfile` vs `subprocess.run(["tar", "xvf"])`. I just hacked the same benchmark to check `tarfile.getmembers()` (returns a list of all tarfile members) vs running `tar -tf` in a subprocess. The results were inconclusive on the small set of relatively small tarballs in the functional test suite, so I went looking for a controller with a range of sizes on the production server:

```
Starting uperf_rhel9.2_5.14.0.284.6.1.el9_intel_10gb_2023.03.31T15.43.12 (27.6 MB)
	 average tarfile list time: 4.71642
	 average tar list time: 3.98687
Starting uperf_rhel9.2_5.14.0-284.6.1.el9_intel_25gb_2023.03.31T20.04.12 (30.9 MB)
	 average tarfile list time: 4.75106
	 average tar list time: 4.02591
Starting uperf_rhel9.2_5.14.0-9.1.el9_intel_40gb_2023.04.06T04.20.35 (31.5 MB)
	 average tarfile list time: 4.81098
	 average tar list time: 4.07516
Starting uperf_rhel9.2_5.14.0.9.1.el9_intel_10gb_jumbo_2023.04.06T15.11.52 (30.8 MB)
	 average tarfile list time: 4.73134
	 average tar list time: 4.00613
Starting uperf_rhel9.2_5.14.0-9.1.el9_mellanox_100gb_2023.04.06T08.41.22 (44.1 MB)
	 average tarfile list time: 7.3921
	 average tar list time: 6.30373
Starting uperf_rhel88_4.18.0-4.18.0-477.3.1_kernel_intel_xl710_40gb_jumbo_2023.03.25T22.18.01 (2.4 GB)
	 average tarfile list time: 43.406
	 average tar list time: 32.6225
Starting uperf_rhel88_4.18.0-4.18.0-477.3.1_kernel_mellanox_cx5_100gb_jumbo_2023.03.26T03.19.50 (3.8 GB)
	 average tarfile list time: 65.4485
	 average tar list time: 48.7486
Starting uperf_rhel9.2_5.14.0.284.el9_intel_10gb_2023.03.21T15.58.10 (30.8 MB)
	 average tarfile list time: 4.74181
	 average tar list time: 3.99639
```

As the files get larger, (and we have a *lot* of very large tarballs), the advantage of spawning the `tar` command becomes fairly significant.

Reply:
Thanks for confirming that, Dave.

I don't know whether to be impressed by the `tar(1)` utility (I mean, I knew it was good...) or disappointed in the Python standard library (well, Python is not known for is performance...).

The truth of the situation is that tar is a ridiculous format for storing random access files in, so it would be best if we only touched the tarball once (and best try not to get stuck to it).